app/models/tariff_date.rb

```diff
@@ -1,59 +1,50 @@
 require 'active_model'
 
-class TariffDate
-  include ActiveModel::Validations
-  include ActiveModel::Conversion
-
+class TariffDate < SimpleDelegator
   DATE_KEYS = %w[year month day].freeze
 
-  delegate :day, :month, :year, :to_formatted_s, :strftime, :today?, :>=, :<=, :==, to: :date
+  class << self
+    def build(as_of, attributes)
+      date = if as_of.present?
+               handle_as_of(as_of)
+             else
+               handle_date_params(attributes)
+             end
```

Comment:
It looks like these are two different builders mixed into a unique `build`.

What is unclear to me is the logic, why if we pass in `as_of` them we must rescue with TariffUpdate.latest_applied_import_date,
and 
with attributes we rescue with`Time.zone.today`.

`as_of` and `attributes` are both dates, it sounds to me we'd need two builders:

`build_by_<what_as_of_represent>`
and 
`build_by_<what_the_attributes_represent>`

Reply:
So my motivation was to move the functionality for differently constructed TariffDates behind a single build method. This simplified the api significantly. I can separate these quite easily with two separate build methods but I'd need to validate how the search model drives the different builders depending on the availability of the as_of. This implementation is actually simpler to test

Reply:
I don't know why BZ used `TariffUpdate.latest_applied_import_date` I'm just preserving existing behaviour